docs/web/docs/explanations/architect_api.md

```diff
@@ -0,0 +1,108 @@
+2. The system should include failsafes to allow for retries.
+3. We should prioritize low latency where possible.
+4. The interaction layer should be simple. 
+
+This document aims to describe how we got from these requirements and goals to the system we have now.
+
+## Requirements
+
+### `Worker`/`Agent` registration
+Mephisto allows for multiple stages of filtering for a worker, which leads the registration process to be staged. The complete order is listed below:
+0. `CrowdProvider`-set qualifications may prevent a worker from even seeing a task.
+1. `Qualification`s may be set that a `CrowdProvider` is not aware of. Workers not meeting these qualifications should be filtered.
+2. Configuration requirements may be set, like `maximum_units_per_worker` or `allowed_concurrent` which could prevent a worker from completing more than a maximum number of units on a task or work on too many tasks at once respectively.
+3. Of all available `Unit`s on `Assignment`s the worker hasn't yet worked on, workers may be further filtered by the user-provided `worker_can_do_unit` function.
+
+This also provides a template of the failure conditions for a worker to be made aware of:
+1. `Worker` doesn't meet the qualifications for a task.
+2. `Worker` is working on too many tasks at a time.
+3. `Worker` has completed the maximum amount of units for the given task, as set by the requester.
+4. None of the _currently available_ units are available for `Worker` to access.
+
+### Frontend access to backend `Unit` data
+Mephisto must provide users with two key ways to provide data to a worker:
+1. Setting `AssignmentData` for an `Assignment`, which define the data made available for a `Unit` on the frontend, and which may also be used to duplicate units.
+2. Providing additional data _during_ a live task, including derived data from partial work. These should be both available via pull and push mechanisms.
+
+Covering these two areas ensures that it's possible to create a broad variety of task types.
+
+### Task data submission
+Like backend data, completed task data should be able to be either tracked during the course of a task or submitted at the end, or both. This leads to two main event types:
+1. Posting completed task data for a task during one of the completion points (onboarding, completing main `Unit` content)
+2. Sending arbitrary task data at any point during a task (for data that needs a response, or longer forms of logging)
+
+### Worker Liveliness and Status
+Over the course of a task, there are a few key states to consider:
+1. Connecting
+2. Onboarding
+3. Waiting (for other `Unit`s in an assignment)
+4. In-Task
+5. Completed (task complete, or partner disconnects)
+6. Disconnected (Server disconnect, timeout, return)
+7. Failed to connect (no available tasks)
+
+We've also considered a "post-task" state after the completion of a task for surveys or related content.
```

Comment:
Wonder if another way of reframing this is simply as a "live" task, e.g. #2 under 'Task data submission' above.

That is to say, we just send the final data back as "arbitrary task data", and then the post-task data could also be sent as either further "arbitrary task data" or completion data.

Reply:
The difficulty here is, under the current setup, "live tasks" with multiple workers consider disconnect events of partners to be the end of a task. As such a live task with multiple workers can't be fully submitted until everyone is done, given the current architecture. 

I'd love to take on an alternative, but I don't quite know what one would be.

Reply:
Ah - interesting caveat!